// cliparser/lib/Parser.js
// Generated by CoffeeScript 1.3.3
(function() {
  var ERRORS, ErrorReporter, Parser,
    __slice = [].slice,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Parser = (function() {

    Parser._activeParser = null;

    Parser._lastArgv = null;

    Parser.parse = function(arg, opts) {
      if (opts == null) {
        opts = {};
      }
      if (!(arg != null)) {
        throw ErrorReporter.generate(1);
      }
      if ((this._activeParser != null) && this._lastArgv === arg) {
        return this._activeParser.results;
      } else {
        return this.reparse(arg, opts);
      }
    };

    Parser.reparse = function(arg, opts) {
      if (opts == null) {
        opts = {};
      }
      if (!(arg != null)) {
        throw ErrorReporter.generate(1);
      }
      this._activeParser = new Parser(arg, opts);
      return this._activeParser.results;
    };

    function Parser(args, opts) {
      if (opts == null) {
        opts = {};
      }
      if (!(args != null)) {
        throw ErrorReporter.generate(2);
      }
      if (args.substr != null) {
        args = args.split(" ");
      }
      this.s = [];
      this.ds = {};
      this.dds = {};
      this.rawArgs = args;
      this.results = {};
      this.symlinkOptions = opts;
      this._parse();
    }

    Parser.prototype._parse = function(args) {
      var index, link, results, run, walk, what, _equal, _parser, _ref;
      if (args == null) {
        args = this.rawArgs;
      }
      if (args !== this.rawArgs) {
        this.rawArgs = args;
      }
      index = 0;
      while ((this.rawArgs[index] != null) && this.rawArgs[index][0] !== "-") {
        this.s.push(this.rawArgs[index++]);
      }
      walk = function(from) {
        var list;
        list = [];
        while ((this.rawArgs[from] != null) && this.rawArgs[from][0] !== "-") {
          list.push(this.rawArgs[from++]);
        }
        return list;
      };
      _parser = this;
      run = function() {
        var args, what;
        what = arguments[0], args = 2 <= arguments.length ? __slice.call(arguments, 1) : [];
        return what.apply(_parser, args);
      };
      while (this.rawArgs[index] != null) {
        results = run(walk, index + 1);
        if (this.rawArgs[index][1] === "-") {
          this.dds[this.rawArgs[index].substr(2)] = results;
        } else {
          this.ds[this.rawArgs[index].substr(1)] = results;
        }
        index += results.length + 1;
      }
      _equal = function(n1, n2, a1, a2) {
        var argument, _i, _len, _ref;
        if ((a2[n2] != null) && a2[n2].length > 0) {
          _ref = a2[n2];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            argument = _ref[_i];
            a1[n1].push(argument);
          }
        }
        return a2[n2] = a1[n1];
      };
      _ref = this.symlinkOptions;
      for (what in _ref) {
        link = _ref[what];
        if (what[1] === "-" && (this.dds[what.substr(2)] != null)) {
          _equal(what.substr(2), link.substr(1), this.dds, this.ds);
        } else if (this.ds[what.substr(1)] != null) {
          _equal(what.substr(1), link.substr(2), this.ds, this.dds);
        }
      }
      return this.results = {
        single: this.s,
        dash: this.ds,
        doubledash: this.dds
      };
    };

    return Parser;

  })();

  module.exports = Parser;

  ErrorReporter = (function(_super) {

    __extends(ErrorReporter, _super);

    ErrorReporter.generate = function(errorCode) {
      if (ERRORS[errorCode] != null) {
        return new ErrorReporter(errorCode);
      } else {
        return new ErrorReporter(0);
      }
    };

    function ErrorReporter(errorCode) {
      if (errorCode) {
        this.name = "ParserError";
      } else {
        this.name = "UnknownError";
      }
      this.message = ERRORS[errorCode];
      this.errorCode = errorCode;
    }

    return ErrorReporter;

  })(TypeError);

  ERRORS = ["There is no error to be handled by the Parser O.o", "No Arguments have been supplied in the factory", "No Arguments have been supplied in the parser object"];

}).call(this);
