// isf/lib/isf.js
(function(module){
	/*!
Math.uuid.js (v1.4)
http://www.broofa.com
mailto:robert@broofa.com

Copyright (c) 2010 Robert Kieffer
Dual licensed under the MIT and GPL licenses.
*/

/*
 * Generate a random uuid.
 *
 * USAGE: Math.uuid(length, radix)
 *   length - the desired number of characters
 *   radix  - the number of allowable values for each character.
 *
 * EXAMPLES:
 *   // No arguments  - returns RFC4122, version 4 ID
 *   >>> Math.uuid()
 *   "92329D39-6F5C-4520-ABFC-AAB64544E172"
 *
 *   // One argument - returns ID of the specified length
 *   >>> Math.uuid(15)     // 15 character ID (default base=62)
 *   "VcydxgltxrVZSTV"
 *
 *   // Two arguments - returns ID of the specified length, and radix. (Radix must be <= 62)
 *   >>> Math.uuid(8, 2)  // 8 character ID (base=2)
 *   "01001010"
 *   >>> Math.uuid(8, 10) // 8 character ID (base=10)
 *   "47473046"
 *   >>> Math.uuid(8, 16) // 8 character ID (base=16)
 *   "098F4D35"
 */
(function() {
  // Private array of chars to use
  var CHARS = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'.split('');

  Math.uuid = function (len, radix) {
    var chars = CHARS, uuid = [], i;
    radix = radix || chars.length;

    if (len) {
      // Compact form
      for (i = 0; i < len; i++) uuid[i] = chars[0 | Math.random()*radix];
    } else {
      // rfc4122, version 4 form
      var r;

      // rfc4122 requires these characters
      uuid[8] = uuid[13] = uuid[18] = uuid[23] = '-';
      uuid[14] = '4';

      // Fill in random data.  At i==19 set the high bits of clock sequence as
      // per rfc4122, sec. 4.1.5
      for (i = 0; i < 36; i++) {
        if (!uuid[i]) {
          r = 0 | Math.random()*16;
          uuid[i] = chars[(i == 19) ? (r & 0x3) | 0x8 : r];
        }
      }
    }

    return uuid.join('');
  };

  // A more performant, but slightly bulkier, RFC4122v4 solution.  We boost performance
  // by minimizing calls to random()
  Math.uuidFast = function() {
    var chars = CHARS, uuid = new Array(36), rnd=0, r;
    for (var i = 0; i < 36; i++) {
      if (i==8 || i==13 ||  i==18 || i==23) {
        uuid[i] = '-';
      } else if (i==14) {
        uuid[i] = '4';
      } else {
        if (rnd <= 0x02) rnd = 0x2000000 + (Math.random()*0x1000000)|0;
        r = rnd & 0xf;
        rnd = rnd >> 4;
        uuid[i] = chars[(i == 19) ? (r & 0x3) | 0x8 : r];
      }
    }
    return uuid.join('');
  };

  // A more compact, but less performant, RFC4122v4 solution:
  Math.uuidCompact = function() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
      var r = Math.random()*16|0, v = c == 'x' ? r : (r&0x3|0x8);
      return v.toString(16);
    });
  };
})();
(function(/*! Stitch !*/) {
	if (!this.require) {
		var modules = {}, cache = {}, require = function(name, root) {
			var path = expand(root, name), module = cache[path], fn;
			if (module) {
				return module.exports;
			} else if (fn = modules[path] || modules[path = expand(path, './index')]) {
				module = {id: path, exports: {}};
				try {
					cache[path] = module;
					fn(module.exports, function(name) {
						return require(name, dirname(path));
					}, module);
					return module.exports;
				} catch (err) {
					delete cache[path];
					throw err;
				}
			} else {
				throw 'module \'' + name + '\' not found';
			}
		}, expand = function(root, name) {
			var results = [], parts, part;
			if (/^\.\.?(\/|$)/.test(name)) {
				parts = [root, name].join('/').split('/');
			} else {
				parts = name.split('/');
			}
			for (var i = 0, length = parts.length; i < length; i++) {
				part = parts[i];
				if (part == '..') {
					results.pop();
				} else if (part != '.' && part != '') {
					results.push(part);
				}
			}
			return results.join('/');
		}, dirname = function(path) {
			return path.split('/').slice(0, -1).join('/');
		};
		this.require = function(name) {
			return require(name, '');
		}
		this.require.define = function(bundle) {
			for (var key in bundle)
				modules[key] = bundle[key];
		};
	}
	return this.require.define;
}).call(this)({"Enum": function(exports, require, module) {(function() {
  var Enum;

  Enum = (function() {
    function Enum(items, offset) {
      var item, key, _i, _len;
      if (offset == null) {
        offset = 0;
      }
      for (key = _i = 0, _len = items.length; _i < _len; key = ++_i) {
        item = items[key];
        this[item] = key + offset;
      }
    }

    return Enum;

  })();

  module.exports = Enum;

}).call(this);
}, "ErrorReporter": function(exports, require, module) {(function() {
  var ErrorReporter,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  ErrorReporter = (function() {
    function ErrorReporter() {
      this.toString = __bind(this.toString, this);
    }

    ErrorReporter._errors = {
      "Unknown Error": ["An unknown error has occurred"]
    };

    ErrorReporter._indices = [ErrorReporter._errors["Unknown Error"][0]];

    ErrorReporter._groups = ["Unknown Error"];

    ErrorReporter.wrapCustomError = function(error) {
      return "[" + error.name + "] " + error.message;
    };

    ErrorReporter.generate = function(errorCode, extra) {
      if (extra == null) {
        extra = null;
      }
      return (new this).generate(errorCode, extra);
    };

    ErrorReporter.extended = function() {
      var error, errors, group, key, _i, _len, _ref;
      _ref = this.errors;
      for (group in _ref) {
        errors = _ref[group];
        this._errors[group] = errors;
        this._groups.push(group);
        for (key = _i = 0, _len = errors.length; _i < _len; key = ++_i) {
          error = errors[key];
          this._indices.push(this._errors[group][key]);
        }
      }
      this.prototype._ = this;
      delete this.errors;
      return this.include(ErrorReporter.prototype);
    };

    ErrorReporter.prototype.generate = function(errCode, extra) {
      var errors, group, _ref, _ref1;
      this.errCode = errCode;
      if (extra == null) {
        extra = null;
      }
      if (!this._._indices[this.errCode]) {
        this.name = this._._groups[0];
        this.message = this._._errors[this._._groups[0]][0];
      } else {
        this.message = this._._indices[this.errCode];
        if (extra) {
          this.message += " - Extra Data : " + extra;
        }
        _ref = this._._errors;
        for (group in _ref) {
          errors = _ref[group];
          if (!(_ref1 = this.message, __indexOf.call(errors, _ref1) >= 0)) {
            continue;
          }
          this.name = group;
          break;
        }
      }
      return this;
    };

    ErrorReporter.prototype.toString = function() {
      return "[" + this.name + "] " + this.message + " |" + this.errCode + "|";
    };

    return ErrorReporter;

  })();

  module.exports = ErrorReporter;

}).call(this);
}, "Modules/Mediator": function(exports, require, module) {(function() {
  var Modules;

  Modules = {
    Observer: require("Modules/Observer")
  };

  Modules.Mediator = (function() {
    var extended, included, installTo, key, value, _ref;

    function Mediator() {}

    _ref = Modules.Observer;
    for (key in _ref) {
      value = _ref[key];
      Mediator.prototype[key] = value;
    }

    installTo = function(object) {
      this.delegate("publish", object);
      return this.delegate("subscribe", object);
    };

    included = function() {
      this.prototype.queue = {};
      return this.prototype._delegates = {
        publish: true,
        subscribe: true
      };
    };

    extended = function() {
      this.queue = {};
      return this._delegates = {
        publish: true,
        subscribe: true
      };
    };

    return Mediator;

  })();

  module.exports = Modules.Mediator.prototype;

}).call(this);
}, "Modules/ORM": function(exports, require, module) {(function() {
  var Modules, V,
    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  Modules = {};

  V = require("Variable");

  Modules.ORM = (function() {
    function ORM() {}

    ORM.prototype._identifier = "BasicORM";

    ORM.prototype._reccords = {};

    ORM.prototype._symlinks = {};

    ORM.prototype._head = 0;

    ORM.prototype._props = [];

    ORM.prototype.get = function(which) {
      if (typeof which === "object") {
        return this.getAdv(which);
      }
      return this._symlinks[which] || this._reccords[which] || null;
    };

    ORM.prototype.getAdv = function(what) {
      var check, key, rec, results, _ref, _ref1;
      results = [];
      check = function(rec) {
        var final, k, mod, modfinal, recs, v, val, value, _i, _len;
        for (k in what) {
          v = what[k];
          final = false;
          if (rec[k] == null) {
            break;
          }
          if ((typeof v) === "object") {
            for (mod in v) {
              val = v[mod];
              modfinal = true;
              switch (mod) {
                case "$gt":
                  if ((rec[k].get()) <= val) {
                    modfinal = false;
                    break;
                  }
                  break;
                case "$gte":
                  if ((rec[k].get()) < val) {
                    modfinal = false;
                    break;
                  }
                  break;
                case "$lt":
                  if ((rec[k].get()) >= val) {
                    modfinal = false;
                    break;
                  }
                  break;
                case "$lte":
                  if ((rec[k].get()) > val) {
                    modfinal = false;
                    break;
                  }
                  break;
                case "$contains":
                  recs = rec[k].get();
                  if (recs.constructor !== Array) {
                    modfinal = false;
                    break;
                  }
                  modfinal = false;
                  for (_i = 0, _len = recs.length; _i < _len; _i++) {
                    value = recs[_i];
                    if (value === val) {
                      modfinal = true;
                      break;
                    }
                  }
              }
              if (modfinal === false) {
                break;
              }
            }
            if (modfinal === true) {
              final = true;
            }
          } else if ((rec[k].get()) === v) {
            final = true;
          } else {
            break;
          }
        }
        if (final) {
          return results.push(rec);
        }
      };
      _ref = this._reccords;
      for (key in _ref) {
        rec = _ref[key];
        check(rec);
      }
      _ref1 = this._symlinks;
      for (key in _ref1) {
        rec = _ref1[key];
        check(rec);
      }
      if (results.length === 0) {
        return null;
      }
      if (results.length === 1) {
        return results[0];
      }
      return results;
    };

    ORM.prototype["delete"] = function(which) {
      var _base, _base1;
      if ((_base = this._reccords)[which] == null) {
        _base[which] = null;
      }
      return (_base1 = this._symlinks)[which] != null ? (_base1 = this._symlinks)[which] : _base1[which] = null;
    };

    ORM.prototype.create = function(id, args) {
      var prop, uuid, _i, _len, _ref;
      if (this._reccords == null) {
        this._reccords = {};
      }
      if (args == null) {
        args = {};
      }
      uuid = id || args._id || this._head;
      if (args._id == null) {
        args._id = uuid;
      }
      uuid = Math.uuidFast(uuid);
      args._uuid = uuid;
      args._fn = this;
      if (typeof this.preCreate === "function") {
        this.preCreate(args);
      }
      this._reccords[uuid] = new this(args);
      this._reccords[uuid]._constructor(args);
      if (typeof this.postCreate === "function") {
        this.postCreate(this._reccords[uuid], args);
      }
      if ((id != null) && id !== this._head) {
        this._symlinks[id] = this._reccords[uuid];
      }
      if (uuid === this._head) {
        this._head++;
      }
      _ref = this._props;
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        prop = _ref[_i];
        this._reccords[uuid][prop] = V.spawn();
      }
      return this._reccords[uuid];
    };

    ORM.prototype.reuse = function(which, args) {
      var rez;
      if (args == null) {
        args = {};
      }
      rez = this.get(which);
      if (rez != null) {
        return rez;
      }
      return this.create(which, args);
    };

    ORM.prototype.addProp = function(prop) {
      var key, rec, _ref, _results;
      this._props.push(prop);
      _ref = this._reccords;
      _results = [];
      for (key in _ref) {
        rec = _ref[key];
        _results.push(rec[prop] != null ? rec[prop] : rec[prop] = V.spawn());
      }
      return _results;
    };

    ORM.prototype.removeProp = function(prop) {
      var k, key, p, rec, _i, _len, _ref, _ref1;
      _ref = this._reccords;
      for (key in _ref) {
        rec = _ref[key];
        if (rec[prop] == null) {
          rec[prop] = null;
        }
      }
      _ref1 = this._props;
      for (k = _i = 0, _len = _ref1.length; _i < _len; k = ++_i) {
        p = _ref1[k];
        if (p === prop) {
          return this._props.splice(k, 1);
        }
      }
    };

    ORM.prototype.extended = function() {
      this._excludes = ["_fn", "_uuid", "_id"];
      return this.include({
        _constructor: function(args) {
          var k, key, v, value, valueSet, _results;
          valueSet = {};
          this._uuid = args._uuid || null;
          this._id = args._id || null;
          this.fn = args._fn;
          for (key in args) {
            value = args[key];
            if (__indexOf.call(this.fn._excludes, key) < 0 && (this.constructFilter(key, value)) !== false) {
              valueSet[key] = value;
            }
          }
          if (this.init != null) {
            return this.init.call(this, valueSet);
          }
          _results = [];
          for (k in valueSet) {
            v = valueSet[k];
            _results.push(this[k] = v);
          }
          return _results;
        },
        constructFilter: function(key, value) {
          return true;
        },
        remove: function() {
          return this.parent.remove(this.id);
        }
      });
    };

    return ORM;

  })();

  module.exports = Modules.ORM.prototype;

}).call(this);
}, "Modules/Observer": function(exports, require, module) {(function() {
  var Modules,
    __slice = [].slice;

  Modules = {};

  Modules.Observer = (function() {
    function Observer() {}

    Observer.prototype.delegateEvent = function(event, handler, object) {
      var c, _base;
      if (object == null) {
        object = window;
      }
      if ((event.substr(0, 2)) === "on") {
        event = event.substr(2);
      }
      if ((_base = this.queue)[event] == null) {
        _base[event] = [];
      }
      c = this.queue[event].length;
      this.queue[event].unshift(function() {
        return handler.apply(object, arguments);
      });
      return c;
    };

    Observer.prototype.subscribe = function(event, handler) {
      return this.delegateEvent(event, handler, this);
    };

    Observer.prototype.publish = function() {
      var args, event, handler, key, _ref;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      event = args[0];
      args = args.splice(1);
      if (!event || (this.queue[event] == null)) {
        return this;
      }
      _ref = this.queue[event];
      for (key in _ref) {
        handler = _ref[key];
        if (key !== "__head") {
          handler.apply(this, args);
        }
      }
      return this;
    };

    Observer.prototype.unsubscribe = function(event, id) {
      if (!this.queue[event]) {
        return null;
      }
      if (!this.queue[event][id]) {
        return null;
      }
      return this.queue[event].splice(id, 1);
    };

    Observer.prototype.included = function() {
      return this.prototype.queue = {};
    };

    Observer.prototype.extended = function() {
      return this.queue = {};
    };

    return Observer;

  })();

  module.exports = Modules.Observer.prototype;

}).call(this);
}, "Modules/Overload": function(exports, require, module) {(function() {
  var CRITERIA, Include, Modules, _count,
    __slice = [].slice,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  Modules = {};

  _count = function(object) {
    var key, nr, value;
    nr = 0;
    for (key in object) {
      value = object[key];
      nr++;
    }
    return nr;
  };

  CRITERIA = {
    args: function(crit, args) {
      return args.length === crit;
    }
  };

  Include = (function() {
    function Include() {}

    Include.prototype.overload = function(sets) {
      var helper;
      helper = new Modules.Overload(sets, this);
      return function() {
        var args;
        args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
        helper.parent = this;
        return helper.verifyAll.apply(helper, args);
      };
    };

    return Include;

  })();

  Modules.Overload = (function() {
    function Overload(sets, parent) {
      var aux, i, j, name, set, _i, _j, _ref, _ref1, _ref2;
      this.parent = parent;
      this.verify = __bind(this.verify, this);
      this.verifyAll = __bind(this.verifyAll, this);
      this.names = [];
      this.verifies = [];
      this.handles = [];
      for (name in sets) {
        set = sets[name];
        this.names.push(name);
        this.verifies.push(set["if"] || null);
        this.handles.push(set.then || null);
      }
      for (i = _i = 0, _ref = this.verifies.length - 1; 0 <= _ref ? _i <= _ref : _i >= _ref; i = 0 <= _ref ? ++_i : --_i) {
        for (j = _j = _ref1 = i + 1, _ref2 = this.verifies.length; _ref1 <= _ref2 ? _j <= _ref2 : _j >= _ref2; j = _ref1 <= _ref2 ? ++_j : --_j) {
          if (_count(this.verifies[i]) < _count(this.verifies[j])) {
            aux = this.verifies[i];
            this.verifies[i] = this.verifies[j];
            this.verifies[j] = aux;
            aux = this.names[i];
            this.names[i] = this.names[j];
            this.names[j] = aux;
            aux = this.handles[i];
            this.handles[i] = this.handles[j];
            this.handles[j] = aux;
          }
        }
      }
    }

    Overload.prototype.verifyAll = function() {
      var args, how, key, set, what, _i, _len, _ref;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      this.args = args;
      _ref = this.verifies;
      for (key = _i = 0, _len = _ref.length; _i < _len; key = ++_i) {
        set = _ref[key];
        if (set != null) {
          for (what in set) {
            how = set[what];
            if (!this.verify(what, how)) {
              break;
            }
            return this.handles[key].apply(this.parent, this.args);
          }
        }
      }
      return (this.handles["default"] || this.handles[key - 1]).apply(this.parent, this.args);
    };

    Overload.prototype.verify = function(what, how) {
      if (CRITERIA[what]) {
        return CRITERIA[what](how, this.args);
      } else {
        what = parseInt(what.replace("arg", "")) - 1;
        if (this.args[what] != null) {
          return how.apply(this.parent, this.args);
        }
        return false;
      }
    };

    return Overload;

  })();

  module.exports = Include.prototype;

}).call(this);
}, "Modules/Pythonize": function(exports, require, module) {(function() {
  var CRITERIA, Include, Modules, _count,
    __slice = [].slice,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  Modules = {};

  _count = function(object) {
    var key, nr, value;
    nr = 0;
    for (key in object) {
      value = object[key];
      nr++;
    }
    return nr;
  };

  CRITERIA = {
    args: function(crit, args) {
      return args.length === crit;
    }
  };

  Include = (function() {
    function Include() {}

    Include.prototype.parameterize = function(sets, callback) {
      var helper;
      helper = new Modules.Pythonize(sets, callback);
      return function() {
        var args;
        args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
        helper.parent = this;
        return helper.verifyAll.apply(helper, args);
      };
    };

    return Include;

  })();

  Modules.Pythonize = (function() {
    function Pythonize(sets, callback) {
      var item, newItem, _i, _len;
      this.callback = callback;
      this.verifyAll = __bind(this.verifyAll, this);
      this.parent = null;
      this._options = [];
      for (_i = 0, _len = sets.length; _i < _len; _i++) {
        item = sets[_i];
        newItem = {
          name: item.name || item.toString(),
          "default": item["default"] || null
        };
        this._options.push(newItem);
      }
    }

    Pythonize.prototype.verifyAll = function() {
      var arg, args, curArg, i, items, lastarg, len, _i, _ref, _ref1, _ref2;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      this.args = args;
      this.options = {};
      len = this.args.length - 1;
      i = 0;
      while (this.args.length > 1) {
        curArg = this._options[i];
        arg = this.args.shift();
        this.options[curArg.name] = arg || curArg["default"];
        i++;
      }
      lastarg = this.args.pop();
      items = this.verifyObject(lastarg, len);
      if (len < this._options.length - 1) {
        for (i = _i = _ref = len + (items.length === 0), _ref1 = this._options.length - 1; _ref <= _ref1 ? _i <= _ref1 : _i >= _ref1; i = _ref <= _ref1 ? ++_i : --_i) {
          if (!(_ref2 = this._options[i].name, __indexOf.call(items, _ref2) >= 0)) {
            this.options[this._options[i].name] = this._options[i]["default"];
          }
        }
      }
      return this.callback.apply(this.parent, [this.options]);
    };

    Pythonize.prototype.verifyObject = function(obj, id) {
      var name, omits, option, valid, value, _i, _len, _ref;
      omits = [];
      if (typeof obj === "object") {
        for (name in obj) {
          value = obj[name];
          valid = false;
          _ref = this._options;
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            option = _ref[_i];
            if (option.name === name) {
              valid = true;
              break;
            }
          }
          if (!valid) {
            this.options[this._options[id].name] = obj;
            return [];
          } else {
            omits.push(name);
            this.options[name] = value;
          }
        }
      } else {
        this.options[this._options[id].name] = obj;
      }
      return omits;
    };

    return Pythonize;

  })();

  module.exports = Include.prototype;

}).call(this);
}, "Modules/StateMachine": function(exports, require, module) {(function() {
  var Modules,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  Modules = {};

  Modules.StateMachine = (function() {
    function StateMachine() {
      this.delegateContext = __bind(this.delegateContext, this);
    }

    StateMachine.prototype.extended = function() {
      this._contexts = [];
      return this._activeContext = null;
    };

    StateMachine.prototype.included = function() {
      this.prototype._contexts = [];
      return this.prototype._activeContext = null;
    };

    StateMachine.prototype.delegateContext = function(context) {
      var l;
      if (this._find(context)) {
        return null;
      }
      l = this._contexts.length;
      this._contexts[l] = context;
      if (context.activate == null) {
        context.activate = function() {};
      }
      if (context.deactivate == null) {
        context.deactivate = function() {};
      }
      return this;
    };

    StateMachine.prototype.getActiveContextID = function() {
      return this._activeContext;
    };

    StateMachine.prototype.getActiveContext = function() {
      return this._activeContext;
    };

    StateMachine.prototype.getContext = function(context) {
      return this._contexts[context] || null;
    };

    StateMachine.prototype._find = function(con) {
      var key, value, _i, _len, _ref;
      _ref = this._contexts;
      for (value = _i = 0, _len = _ref.length; _i < _len; value = ++_i) {
        key = _ref[value];
        if (con === key) {
          return value;
        }
      }
      return null;
    };

    StateMachine.prototype.activateContext = function(context) {
      var con;
      con = this._find(context);
      if (con == null) {
        return null;
      }
      if (this._activeContext === con) {
        return true;
      }
      this._activeContext = con;
      return context.activate();
    };

    StateMachine.prototype.deactivateContext = function(context) {
      if ((this._find(context)) == null) {
        return null;
      }
      this._activeContext = null;
      return context.deactivate();
    };

    StateMachine.prototype.switchContext = function(context) {
      var con;
      if (context == null) {
        con = this._activeContext + 1;
        if (con === this._contexts.length) {
          con = 0;
        }
      } else {
        con = this._find(context);
        if (con == null) {
          return null;
        }
      }
      this.deactivateContext(this._contexts[this._activeContext]);
      this.activateContext(this._contexts[con]);
      return this._contexts[con];
    };

    return StateMachine;

  })();

  module.exports = Modules.StateMachine.prototype;

}).call(this);
}, "Object": function(exports, require, module) {(function() {
  var $, Obiect, clone, _excludes,
    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; },
    __slice = [].slice;

  _excludes = ["included", "extended"];

  clone = function(obj) {
    var k, o, v;
    o = obj instanceof Array ? [] : {};
    for (k in obj) {
      v = obj[k];
      if ((v != null) && typeof v === "object") {
        o[k] = clone(v);
      } else {
        o[k] = v;
      }
    }
    return o;
  };

  $ = function(what) {
    return $[what] || null;
  };

  Obiect = (function() {
    var extended, included;

    function Obiect() {}

    Obiect.clone = function(obj) {
      if (obj == null) {
        obj = this;
      }
      debugger;
      return (Obiect.proxy(Obiect.include, (Obiect.proxy(Obiect.extend, function() {}))(obj)))(obj.prototype);
    };

    Obiect.extend = function(obj, into) {
      var k, value, _ref;
      if (into == null) {
        into = this;
      }
      obj = clone(obj);
      for (k in obj) {
        value = obj[k];
        if (!((__indexOf.call(_excludes, k) >= 0) || ((obj._excludes != null) && __indexOf.call(obj._excludes, k) >= 0))) {
          if (into[k] != null) {
            if (into["super"] == null) {
              into["super"] = {};
            }
            into["super"][k] = into[k];
          }
          into[k] = value;
        }
      }
      if ((_ref = obj.extended) != null) {
        _ref.call(into);
      }
      return this;
    };

    Obiect.include = function(obj, into) {
      var key, value, _ref;
      if (into == null) {
        into = this;
      }
      obj = clone(obj);
      for (key in obj) {
        value = obj[key];
        into.prototype[key] = value;
      }
      if ((_ref = obj.included) != null) {
        _ref.call(into);
      }
      return this;
    };

    Obiect.proxy = function() {
      var to, what,
        _this = this;
      what = arguments[0];
      to = arguments[1];
      if (typeof what === "function") {
        return function() {
          var args;
          args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
          return what.apply(to, args);
        };
      } else {
        return this[what];
      }
    };

    Obiect.delegate = function(property, context) {
      var _ref;
      if ((((_ref = this._delegates) != null ? _ref[property] : void 0) != null) === false && this._deleagates[property] !== false) {
        trigger("Cannot delegate member " + property + " to " + context);
      }
      return context[property] = this.proxy(function() {
        return this[property](arguments);
      }, this);
    };

    Obiect.echo = function() {
      var args, owner, prefix, _d;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      _d = new Date;
      owner = "<not supported>";
      if (this.__proto__ != null) {
        owner = this.__proto__.constructor.name;
      }
      prefix = "[" + (_d.getHours()) + ":" + (_d.getMinutes()) + ":" + (_d.getSeconds()) + "][" + (this.name || owner) + "]";
      if (args[0] === "") {
        args[0] = prefix;
      } else {
        args[0] = "" + prefix + " " + args[0];
      }
      console.log(args);
      return this;
    };

    Obiect.log = function() {
      var args;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      if ((typeof IS !== "undefined" && IS !== null ? IS.isDev : void 0) || window.isDev || (typeof root !== "undefined" && root !== null ? root.isDev : void 0) || isDev) {
        args.unshift("");
        this.echo.apply(this, args);
      }
      return this;
    };

    extended = function() {};

    included = function() {};

    Obiect.include({
      proxy: Obiect.proxy,
      log: Obiect.log,
      echo: Obiect.echo
    });

    return Obiect;

  })();

  module.exports = Obiect;

}).call(this);
}, "Promise": function(exports, require, module) {(function() {
  var Promise,
    __slice = [].slice;

  Promise = (function() {
    function Promise(promise) {
      if (promise instanceof Promise) {
        return promise;
      }
      this.callbacks = [];
    }

    Promise.prototype.then = function(ok, err, progr) {
      this.callbacks.push({
        ok: ok,
        error: err,
        progress: progr
      });
      return this;
    };

    Promise.prototype.resolve = function() {
      var args, callback, time,
        _this = this;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      callback = this.callbacks.shift();
      if (callback && callback.ok) {
        callback.ok.apply(this, args);
      } else {
        time = setTimeout(function() {
          clearTimeout(time);
          return _this.resolve.apply(_this, args);
        }, 50);
      }
      return this;
    };

    Promise.prototype.reject = function() {
      var args, callback, time,
        _this = this;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      callback = this.callbacks.shift();
      if (callback && callback.error) {
        callback.error.apply(this, args);
      } else {
        time = setTimeout(function() {
          clearTimeout(time);
          return _this.reject.apply(_this, args);
        }, 50);
      }
      return this;
    };

    Promise.prototype.progress = function() {
      var args, callback;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      callback = this.callbacks[0];
      if (callback && callback.progress) {
        callback.progress.apply(this, args);
      }
      return this;
    };

    return Promise;

  })();

  module.exports = Promise;

}).call(this);
}, "Variable": function(exports, require, module) {(function() {
  var Variable, _ref,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Variable = (function(_super) {
    __extends(Variable, _super);

    function Variable() {
      _ref = Variable.__super__.constructor.apply(this, arguments);
      return _ref;
    }

    Variable.spawn = function() {
      var x;
      x = new this;
      x._value = null;
      return x;
    };

    Variable.prototype.get = function() {
      return this._value;
    };

    Variable.prototype.set = function(value) {
      return this._value = value;
    };

    Variable.prototype.add = function(reccord) {
      if ((this._value == null) || this._value.constructor !== Array) {
        this._value = [];
      }
      return this._value.push(reccord);
    };

    return Variable;

  })(require("Object"));

  if (typeof module !== "undefined" && module !== null) {
    module.exports = Variable;
  }

}).call(this);
}, "async": function(exports, require, module) {/*global setImmediate: false, setTimeout: false, console: false */
(function () {

    var async = {};

    // global on the server, window in the browser
    var root, previous_async;

    root = this;
    if (root != null) {
      previous_async = root.async;
    }

    async.noConflict = function () {
        root.async = previous_async;
        return async;
    };

    function only_once(fn) {
        var called = false;
        return function() {
            if (called) throw new Error("Callback was already called.");
            called = true;
            fn.apply(root, arguments);
        }
    }

    //// cross-browser compatiblity functions ////

    var _each = function (arr, iterator) {
        if (arr.forEach) {
            return arr.forEach(iterator);
        }
        for (var i = 0; i < arr.length; i += 1) {
            iterator(arr[i], i, arr);
        }
    };

    var _map = function (arr, iterator) {
        if (arr.map) {
            return arr.map(iterator);
        }
        var results = [];
        _each(arr, function (x, i, a) {
            results.push(iterator(x, i, a));
        });
        return results;
    };

    var _reduce = function (arr, iterator, memo) {
        if (arr.reduce) {
            return arr.reduce(iterator, memo);
        }
        _each(arr, function (x, i, a) {
            memo = iterator(memo, x, i, a);
        });
        return memo;
    };

    var _keys = function (obj) {
        if (Object.keys) {
            return Object.keys(obj);
        }
        var keys = [];
        for (var k in obj) {
            if (obj.hasOwnProperty(k)) {
                keys.push(k);
            }
        }
        return keys;
    };

    //// exported async module functions ////

    //// nextTick implementation with browser-compatible fallback ////
    if (typeof process === 'undefined' || !(process.nextTick)) {
        if (typeof setImmediate === 'function') {
            async.nextTick = function (fn) {
                // not a direct alias for IE10 compatibility
                setImmediate(fn);
            };
            async.setImmediate = async.nextTick;
        }
        else {
            async.nextTick = function (fn) {
                setTimeout(fn, 0);
            };
            async.setImmediate = async.nextTick;
        }
    }
    else {
        async.nextTick = process.nextTick;
        if (typeof setImmediate !== 'undefined') {
            async.setImmediate = setImmediate;
        }
        else {
            async.setImmediate = async.nextTick;
        }
    }

    async.each = function (arr, iterator, callback) {
        callback = callback || function () {};
        if (!arr.length) {
            return callback();
        }
        var completed = 0;
        _each(arr, function (x) {
            iterator(x, only_once(function (err) {
                if (err) {
                    callback(err);
                    callback = function () {};
                }
                else {
                    completed += 1;
                    if (completed >= arr.length) {
                        callback(null);
                    }
                }
            }));
        });
    };
    async.forEach = async.each;

    async.eachSeries = function (arr, iterator, callback) {
        callback = callback || function () {};
        if (!arr.length) {
            return callback();
        }
        var completed = 0;
        var iterate = function () {
            iterator(arr[completed], function (err) {
                if (err) {
                    callback(err);
                    callback = function () {};
                }
                else {
                    completed += 1;
                    if (completed >= arr.length) {
                        callback(null);
                    }
                    else {
                        iterate();
                    }
                }
            });
        };
        iterate();
    };
    async.forEachSeries = async.eachSeries;

    async.eachLimit = function (arr, limit, iterator, callback) {
        var fn = _eachLimit(limit);
        fn.apply(null, [arr, iterator, callback]);
    };
    async.forEachLimit = async.eachLimit;

    var _eachLimit = function (limit) {

        return function (arr, iterator, callback) {
            callback = callback || function () {};
            if (!arr.length || limit <= 0) {
                return callback();
            }
            var completed = 0;
            var started = 0;
            var running = 0;

            (function replenish () {
                if (completed >= arr.length) {
                    return callback();
                }

                while (running < limit && started < arr.length) {
                    started += 1;
                    running += 1;
                    iterator(arr[started - 1], function (err) {
                        if (err) {
                            callback(err);
                            callback = function () {};
                        }
                        else {
                            completed += 1;
                            running -= 1;
                            if (completed >= arr.length) {
                                callback();
                            }
                            else {
                                replenish();
                            }
                        }
                    });
                }
            })();
        };
    };


    var doParallel = function (fn) {
        return function () {
            var args = Array.prototype.slice.call(arguments);
            return fn.apply(null, [async.each].concat(args));
        };
    };
    var doParallelLimit = function(limit, fn) {
        return function () {
            var args = Array.prototype.slice.call(arguments);
            return fn.apply(null, [_eachLimit(limit)].concat(args));
        };
    };
    var doSeries = function (fn) {
        return function () {
            var args = Array.prototype.slice.call(arguments);
            return fn.apply(null, [async.eachSeries].concat(args));
        };
    };


    var _asyncMap = function (eachfn, arr, iterator, callback) {
        var results = [];
        arr = _map(arr, function (x, i) {
            return {index: i, value: x};
        });
        eachfn(arr, function (x, callback) {
            iterator(x.value, function (err, v) {
                results[x.index] = v;
                callback(err);
            });
        }, function (err) {
            callback(err, results);
        });
    };
    async.map = doParallel(_asyncMap);
    async.mapSeries = doSeries(_asyncMap);
    async.mapLimit = function (arr, limit, iterator, callback) {
        return _mapLimit(limit)(arr, iterator, callback);
    };

    var _mapLimit = function(limit) {
        return doParallelLimit(limit, _asyncMap);
    };

    // reduce only has a series version, as doing reduce in parallel won't
    // work in many situations.
    async.reduce = function (arr, memo, iterator, callback) {
        async.eachSeries(arr, function (x, callback) {
            iterator(memo, x, function (err, v) {
                memo = v;
                callback(err);
            });
        }, function (err) {
            callback(err, memo);
        });
    };
    // inject alias
    async.inject = async.reduce;
    // foldl alias
    async.foldl = async.reduce;

    async.reduceRight = function (arr, memo, iterator, callback) {
        var reversed = _map(arr, function (x) {
            return x;
        }).reverse();
        async.reduce(reversed, memo, iterator, callback);
    };
    // foldr alias
    async.foldr = async.reduceRight;

    var _filter = function (eachfn, arr, iterator, callback) {
        var results = [];
        arr = _map(arr, function (x, i) {
            return {index: i, value: x};
        });
        eachfn(arr, function (x, callback) {
            iterator(x.value, function (v) {
                if (v) {
                    results.push(x);
                }
                callback();
            });
        }, function (err) {
            callback(_map(results.sort(function (a, b) {
                return a.index - b.index;
            }), function (x) {
                return x.value;
            }));
        });
    };
    async.filter = doParallel(_filter);
    async.filterSeries = doSeries(_filter);
    // select alias
    async.select = async.filter;
    async.selectSeries = async.filterSeries;

    var _reject = function (eachfn, arr, iterator, callback) {
        var results = [];
        arr = _map(arr, function (x, i) {
            return {index: i, value: x};
        });
        eachfn(arr, function (x, callback) {
            iterator(x.value, function (v) {
                if (!v) {
                    results.push(x);
                }
                callback();
            });
        }, function (err) {
            callback(_map(results.sort(function (a, b) {
                return a.index - b.index;
            }), function (x) {
                return x.value;
            }));
        });
    };
    async.reject = doParallel(_reject);
    async.rejectSeries = doSeries(_reject);

    var _detect = function (eachfn, arr, iterator, main_callback) {
        eachfn(arr, function (x, callback) {
            iterator(x, function (result) {
                if (result) {
                    main_callback(x);
                    main_callback = function () {};
                }
                else {
                    callback();
                }
            });
        }, function (err) {
            main_callback();
        });
    };
    async.detect = doParallel(_detect);
    async.detectSeries = doSeries(_detect);

    async.some = function (arr, iterator, main_callback) {
        async.each(arr, function (x, callback) {
            iterator(x, function (v) {
                if (v) {
                    main_callback(true);
                    main_callback = function () {};
                }
                callback();
            });
        }, function (err) {
            main_callback(false);
        });
    };
    // any alias
    async.any = async.some;

    async.every = function (arr, iterator, main_callback) {
        async.each(arr, function (x, callback) {
            iterator(x, function (v) {
                if (!v) {
                    main_callback(false);
                    main_callback = function () {};
                }
                callback();
            });
        }, function (err) {
            main_callback(true);
        });
    };
    // all alias
    async.all = async.every;

    async.sortBy = function (arr, iterator, callback) {
        async.map(arr, function (x, callback) {
            iterator(x, function (err, criteria) {
                if (err) {
                    callback(err);
                }
                else {
                    callback(null, {value: x, criteria: criteria});
                }
            });
        }, function (err, results) {
            if (err) {
                return callback(err);
            }
            else {
                var fn = function (left, right) {
                    var a = left.criteria, b = right.criteria;
                    return a < b ? -1 : a > b ? 1 : 0;
                };
                callback(null, _map(results.sort(fn), function (x) {
                    return x.value;
                }));
            }
        });
    };

    async.auto = function (tasks, callback) {
        callback = callback || function () {};
        var keys = _keys(tasks);
        if (!keys.length) {
            return callback(null);
        }

        var results = {};

        var listeners = [];
        var addListener = function (fn) {
            listeners.unshift(fn);
        };
        var removeListener = function (fn) {
            for (var i = 0; i < listeners.length; i += 1) {
                if (listeners[i] === fn) {
                    listeners.splice(i, 1);
                    return;
                }
            }
        };
        var taskComplete = function () {
            _each(listeners.slice(0), function (fn) {
                fn();
            });
        };

        addListener(function () {
            if (_keys(results).length === keys.length) {
                callback(null, results);
                callback = function () {};
            }
        });

        _each(keys, function (k) {
            var task = (tasks[k] instanceof Function) ? [tasks[k]]: tasks[k];
            var taskCallback = function (err) {
                var args = Array.prototype.slice.call(arguments, 1);
                if (args.length <= 1) {
                    args = args[0];
                }
                if (err) {
                    var safeResults = {};
                    _each(_keys(results), function(rkey) {
                        safeResults[rkey] = results[rkey];
                    });
                    safeResults[k] = args;
                    callback(err, safeResults);
                    // stop subsequent errors hitting callback multiple times
                    callback = function () {};
                }
                else {
                    results[k] = args;
                    async.setImmediate(taskComplete);
                }
            };
            var requires = task.slice(0, Math.abs(task.length - 1)) || [];
            var ready = function () {
                return _reduce(requires, function (a, x) {
                    return (a && results.hasOwnProperty(x));
                }, true) && !results.hasOwnProperty(k);
            };
            if (ready()) {
                task[task.length - 1](taskCallback, results);
            }
            else {
                var listener = function () {
                    if (ready()) {
                        removeListener(listener);
                        task[task.length - 1](taskCallback, results);
                    }
                };
                addListener(listener);
            }
        });
    };

    async.waterfall = function (tasks, callback) {
        callback = callback || function () {};
        if (tasks.constructor !== Array) {
          var err = new Error('First argument to waterfall must be an array of functions');
          return callback(err);
        }
        if (!tasks.length) {
            return callback();
        }
        var wrapIterator = function (iterator) {
            return function (err) {
                if (err) {
                    callback.apply(null, arguments);
                    callback = function () {};
                }
                else {
                    var args = Array.prototype.slice.call(arguments, 1);
                    var next = iterator.next();
                    if (next) {
                        args.push(wrapIterator(next));
                    }
                    else {
                        args.push(callback);
                    }
                    async.setImmediate(function () {
                        iterator.apply(null, args);
                    });
                }
            };
        };
        wrapIterator(async.iterator(tasks))();
    };

    var _parallel = function(eachfn, tasks, callback) {
        callback = callback || function () {};
        if (tasks.constructor === Array) {
            eachfn.map(tasks, function (fn, callback) {
                if (fn) {
                    fn(function (err) {
                        var args = Array.prototype.slice.call(arguments, 1);
                        if (args.length <= 1) {
                            args = args[0];
                        }
                        callback.call(null, err, args);
                    });
                }
            }, callback);
        }
        else {
            var results = {};
            eachfn.each(_keys(tasks), function (k, callback) {
                tasks[k](function (err) {
                    var args = Array.prototype.slice.call(arguments, 1);
                    if (args.length <= 1) {
                        args = args[0];
                    }
                    results[k] = args;
                    callback(err);
                });
            }, function (err) {
                callback(err, results);
            });
        }
    };

    async.parallel = function (tasks, callback) {
        _parallel({ map: async.map, each: async.each }, tasks, callback);
    };

    async.parallelLimit = function(tasks, limit, callback) {
        _parallel({ map: _mapLimit(limit), each: _eachLimit(limit) }, tasks, callback);
    };

    async.series = function (tasks, callback) {
        callback = callback || function () {};
        if (tasks.constructor === Array) {
            async.mapSeries(tasks, function (fn, callback) {
                if (fn) {
                    fn(function (err) {
                        var args = Array.prototype.slice.call(arguments, 1);
                        if (args.length <= 1) {
                            args = args[0];
                        }
                        callback.call(null, err, args);
                    });
                }
            }, callback);
        }
        else {
            var results = {};
            async.eachSeries(_keys(tasks), function (k, callback) {
                tasks[k](function (err) {
                    var args = Array.prototype.slice.call(arguments, 1);
                    if (args.length <= 1) {
                        args = args[0];
                    }
                    results[k] = args;
                    callback(err);
                });
            }, function (err) {
                callback(err, results);
            });
        }
    };

    async.iterator = function (tasks) {
        var makeCallback = function (index) {
            var fn = function () {
                if (tasks.length) {
                    tasks[index].apply(null, arguments);
                }
                return fn.next();
            };
            fn.next = function () {
                return (index < tasks.length - 1) ? makeCallback(index + 1): null;
            };
            return fn;
        };
        return makeCallback(0);
    };

    async.apply = function (fn) {
        var args = Array.prototype.slice.call(arguments, 1);
        return function () {
            return fn.apply(
                null, args.concat(Array.prototype.slice.call(arguments))
            );
        };
    };

    var _concat = function (eachfn, arr, fn, callback) {
        var r = [];
        eachfn(arr, function (x, cb) {
            fn(x, function (err, y) {
                r = r.concat(y || []);
                cb(err);
            });
        }, function (err) {
            callback(err, r);
        });
    };
    async.concat = doParallel(_concat);
    async.concatSeries = doSeries(_concat);

    async.whilst = function (test, iterator, callback) {
        if (test()) {
            iterator(function (err) {
                if (err) {
                    return callback(err);
                }
                async.whilst(test, iterator, callback);
            });
        }
        else {
            callback();
        }
    };

    async.doWhilst = function (iterator, test, callback) {
        iterator(function (err) {
            if (err) {
                return callback(err);
            }
            if (test()) {
                async.doWhilst(iterator, test, callback);
            }
            else {
                callback();
            }
        });
    };

    async.until = function (test, iterator, callback) {
        if (!test()) {
            iterator(function (err) {
                if (err) {
                    return callback(err);
                }
                async.until(test, iterator, callback);
            });
        }
        else {
            callback();
        }
    };

    async.doUntil = function (iterator, test, callback) {
        iterator(function (err) {
            if (err) {
                return callback(err);
            }
            if (!test()) {
                async.doUntil(iterator, test, callback);
            }
            else {
                callback();
            }
        });
    };

    async.queue = function (worker, concurrency) {
        if (concurrency === undefined) {
            concurrency = 1;
        }
        function _insert(q, data, pos, callback) {
          if(data.constructor !== Array) {
              data = [data];
          }
          _each(data, function(task) {
              var item = {
                  data: task,
                  callback: typeof callback === 'function' ? callback : null
              };

              if (pos) {
                q.tasks.unshift(item);
              } else {
                q.tasks.push(item);
              }

              if (q.saturated && q.tasks.length === concurrency) {
                  q.saturated();
              }
              async.setImmediate(q.process);
          });
        }

        var workers = 0;
        var q = {
            tasks: [],
            concurrency: concurrency,
            saturated: null,
            empty: null,
            drain: null,
            push: function (data, callback) {
              _insert(q, data, false, callback);
            },
            unshift: function (data, callback) {
              _insert(q, data, true, callback);
            },
            process: function () {
                if (workers < q.concurrency && q.tasks.length) {
                    var task = q.tasks.shift();
                    if (q.empty && q.tasks.length === 0) {
                        q.empty();
                    }
                    workers += 1;
                    var next = function () {
                        workers -= 1;
                        if (task.callback) {
                            task.callback.apply(task, arguments);
                        }
                        if (q.drain && q.tasks.length + workers === 0) {
                            q.drain();
                        }
                        q.process();
                    };
                    var cb = only_once(next);
                    worker(task.data, cb);
                }
            },
            length: function () {
                return q.tasks.length;
            },
            running: function () {
                return workers;
            }
        };
        return q;
    };

    async.cargo = function (worker, payload) {
        var working     = false,
            tasks       = [];

        var cargo = {
            tasks: tasks,
            payload: payload,
            saturated: null,
            empty: null,
            drain: null,
            push: function (data, callback) {
                if(data.constructor !== Array) {
                    data = [data];
                }
                _each(data, function(task) {
                    tasks.push({
                        data: task,
                        callback: typeof callback === 'function' ? callback : null
                    });
                    if (cargo.saturated && tasks.length === payload) {
                        cargo.saturated();
                    }
                });
                async.setImmediate(cargo.process);
            },
            process: function process() {
                if (working) return;
                if (tasks.length === 0) {
                    if(cargo.drain) cargo.drain();
                    return;
                }

                var ts = typeof payload === 'number'
                            ? tasks.splice(0, payload)
                            : tasks.splice(0);

                var ds = _map(ts, function (task) {
                    return task.data;
                });

                if(cargo.empty) cargo.empty();
                working = true;
                worker(ds, function () {
                    working = false;

                    var args = arguments;
                    _each(ts, function (data) {
                        if (data.callback) {
                            data.callback.apply(null, args);
                        }
                    });

                    process();
                });
            },
            length: function () {
                return tasks.length;
            },
            running: function () {
                return working;
            }
        };
        return cargo;
    };

    var _console_fn = function (name) {
        return function (fn) {
            var args = Array.prototype.slice.call(arguments, 1);
            fn.apply(null, args.concat([function (err) {
                var args = Array.prototype.slice.call(arguments, 1);
                if (typeof console !== 'undefined') {
                    if (err) {
                        if (console.error) {
                            console.error(err);
                        }
                    }
                    else if (console[name]) {
                        _each(args, function (x) {
                            console[name](x);
                        });
                    }
                }
            }]));
        };
    };
    async.log = _console_fn('log');
    async.dir = _console_fn('dir');
    /*async.info = _console_fn('info');
    async.warn = _console_fn('warn');
    async.error = _console_fn('error');*/

    async.memoize = function (fn, hasher) {
        var memo = {};
        var queues = {};
        hasher = hasher || function (x) {
            return x;
        };
        var memoized = function () {
            var args = Array.prototype.slice.call(arguments);
            var callback = args.pop();
            var key = hasher.apply(null, args);
            if (key in memo) {
                callback.apply(null, memo[key]);
            }
            else if (key in queues) {
                queues[key].push(callback);
            }
            else {
                queues[key] = [callback];
                fn.apply(null, args.concat([function () {
                    memo[key] = arguments;
                    var q = queues[key];
                    delete queues[key];
                    for (var i = 0, l = q.length; i < l; i++) {
                      q[i].apply(null, arguments);
                    }
                }]));
            }
        };
        memoized.memo = memo;
        memoized.unmemoized = fn;
        return memoized;
    };

    async.unmemoize = function (fn) {
      return function () {
        return (fn.unmemoized || fn).apply(null, arguments);
      };
    };

    async.times = function (count, iterator, callback) {
        var counter = [];
        for (var i = 0; i < count; i++) {
            counter.push(i);
        }
        return async.map(counter, iterator, callback);
    };

    async.timesSeries = function (count, iterator, callback) {
        var counter = [];
        for (var i = 0; i < count; i++) {
            counter.push(i);
        }
        return async.mapSeries(counter, iterator, callback);
    };

    async.compose = function (/* functions... */) {
        var fns = Array.prototype.reverse.call(arguments);
        return function () {
            var that = this;
            var args = Array.prototype.slice.call(arguments);
            var callback = args.pop();
            async.reduce(fns, args, function (newargs, fn, cb) {
                fn.apply(that, newargs.concat([function () {
                    var err = arguments[0];
                    var nextargs = Array.prototype.slice.call(arguments, 1);
                    cb(err, nextargs);
                }]))
            },
            function (err, results) {
                callback.apply(that, [err].concat(results));
            });
        };
    };

    var _applyEach = function (eachfn, fns /*args...*/) {
        var go = function () {
            var that = this;
            var args = Array.prototype.slice.call(arguments);
            var callback = args.pop();
            return eachfn(fns, function (fn, cb) {
                fn.apply(that, args.concat([cb]));
            },
            callback);
        };
        if (arguments.length > 2) {
            var args = Array.prototype.slice.call(arguments, 2);
            return go.apply(this, args);
        }
        else {
            return go;
        }
    };
    async.applyEach = doParallel(_applyEach);
    async.applyEachSeries = doSeries(_applyEach);

    async.forever = function (fn, callback) {
        function next(err) {
            if (err) {
                if (callback) {
                    return callback(err);
                }
                throw err;
            }
            fn(next);
        }
        next();
    };

    // AMD / RequireJS
    if (typeof define !== 'undefined' && define.amd) {
        define([], function () {
            return async;
        });
    }
    // Node.js
    else if (typeof module !== 'undefined' && module.exports) {
        module.exports = async;
    }
    // included directly via <script> tag
    else {
        root.async = async;
    }

}());
}});

	var require = this.require;
	(function() {
  var IS;

  require('Object');

  require('async');

  IS = {
    Variable: require('Variable'),
    Promise: require('Promise'),
    Enum: require('Enum'),
    ErrorReporter: require('ErrorReporter'),
    Object: require('Object'),
    Modules: {
      ORM: require('Modules/ORM'),
      Pythonize: require('Modules/Pythonize'),
      Observer: require('Modules/Observer'),
      StateMachine: require('Modules/StateMachine'),
      Mediator: require('Modules/Mediator'),
      Overload: require('Modules/Overload')
    }
  };

  if (typeof window !== "undefined" && window !== null) {
    window.IS = IS;
  }

  if (typeof module !== "undefined" && module !== null) {
    module.exports = IS;
  }

  if (typeof root !== "undefined" && root !== null) {
    root.IS = IS;
  }

}).call(this);

}).call({}, typeof(module) == "undefined" ? (typeof(window) == "undefined" ? root : window) : module);
